### ISSUE TITLE
Allowing Microkit to not require a kernel patch

### ISSUE THREAD

**Issue by Ivan-Velickovic** [seL4/microkit#52]
There is currently a single patch to the seL4 kernel (https://github.com/seL4/seL4/commit/c8ef493d038b81cc43f82c0190788e4b3bdb4d9d) that Microkit depends on.

**What is the patch for?**

The patch allows the boot-loader, which runs before seL4, to specify an extra region of device untyped at run-time when the boot-loader jumps to seL4.

**Why is the patch necessary?**

The boot-loader that Microkit systems use places all binaries in the expected places in physical memory *before* jumping to seL4. This includes binary for seL4 itself, the initial task (aka the monitor), as well as the binary for each protection domain.

However, when seL4 starts, it clears all physical memory that it defines as 'normal'. This means all memory in RAM (except seL4 itself and the initial task) will be cleared. Without the kernel patch, this would include the PD ELFs, meaning that when the PDs start, all the code for them will be gone.

In order to avoid this, we place all the PD ELFs in a contiguous region of physical memory that we tell seL4 at boot is 'device' memory, meaning that it *won't* clear it and hence the PD ELFs remain in tact after seL4 has booted.

**How will the patch be removed?**

Assuming there are no blocking issues, the new Rust capDL initialiser can be used to embed the binaries for each PD in the initialiser itself, while not copying the binaries around in memory. This allows us to have the memory efficiency desired without a patch to the kernel.

After the kernel patch is removed, a specific branch of seL4 will no longer be required. Note that Microkit may be pinned to a specific version of seL4 still, so not just any commit of the kernel will work.

**Comment by Ivan-Velickovic** [comment]
Thanks for your comments.

> To me it seems your alternative solution works fine if you don't need memory regions with known physical addresses (for e.g. U-boot), and don't need contiguous memory for e.g. VMs.

I am a bit confused by this, I don't see how this would stop allocating a large contiguous region of memory? The Microkit tool has only two restrictions on where it can allocate user-specified memory regions from. One is the region for seL4 itself, and the other is for the initial task. Everything else could be in theory used by the user. Given neither seL4 or the initial task use up much space relative to the typical size of RAM, I don't see how allocating a large contiguous chunk of memory for a VM at particular physical address (e.g for 1-1 mappings for DMA) would be difficult. If it would be difficult with the proposed approach, it would be difficult with the existing approach, as both approaches would use about the same amount of memory.

>  This seems harder to do if you embed the image into the loader.

Can you clarify what you mean by "loader" in this case? Loader means boot-loader to me but some other people use it to mean the initial task.

**Comment by Indanz** [comment]
> I am a bit confused by this, I don't see how this would stop us allocating a large contiguous region of memory?

In our case we tried allocating exactly half of 1 GB memory region, but that was impossible to do because of fragmentation and reserved UTs at both start and end of the address space. There were some patches fixing some of those unnecessary fragmentation issues in seL4, but I don't think all were fixed yet. So the only way to get this was reserving that area as device memory in DT.

Your Microkit patch would allow the same to be achieved, but you can't embed half the RAM in the initialiser itself. Or can you tell the initialiser where to put itself and what memory to reserve? If so, perhaps it's still possible.

> Can you clarify what you mean by "loader" in this case? Loader means boot-loader to me but some other people use it to mean the initial task.

I meant it generically. If whatever any loader needs to load is embedded into it, then reloading an updated binary will always be more difficult, because you can't reuse the existing loading code. Except if you can reserve extra space and overwrite the embedded binary, but you have to be very careful when doing that.

**Comment by Indanz** [comment]
> a) stops the kernel from using large pages for the initial task

True, but that still doesn't give the user control whether it does or not.

> b) nothing stops the microkit loader from being written in a way such that the initial task frames get recycled (previously the loader completely remapped it's virtual address space and pages; this is not done simply because it's simpler).

That only solves the permission problem.

> Sure, it's inflexible, but I don't know how you'd make it any better without creating some very generic and complicated boot code so that seL4 knows how to allocate things, because there's a lot of flexibility that could be desired. It's nicer (imo) to do all of this in the initial task - that's what it's there for. I don't see how anything more complicated could be less of a hack.
> 
> (I will note that I think the way that DTBs are passed is also a hack and should have just been part of the initial task contiguous region).

Exactly this! Microkit could have passed all extra data as a "DTB", and it would have worked. Only cost would be one memcpy.

So the better generic solution is replacing the DTB arguments with a generic device region thing that the Microkit patch did.

But I'd go one step further and instead of using a DTB for seL4 at all, let the loader tell seL4 what normal memory there is it can use. And the seL4 kernel should understand and load ELF files properly (having specific restrictions to simplify this is fine of course). Parsing ELF program headers isn't very complicated for static programs (if you ignore thread-local storage, that can be pushed to user space).

> > The seL4 kernel will map those extra pages unnecessarily at bootup with RWX permissions inside the initial's thread address space, that's not really what you want.
> 
> Sure - but like I said we could do some extra invocations to remap these or unmap them entirely from the initial task.

If that's gigabytes, you slow down boot even more. Performance wise the damage has been done already.

Then again, if you want to reduce the attack surface of the initial task you do need to unmap and remap frames anyway.

> (But at the same time the initial task has all the capabilities so it has to be trusted anyway; and in the microkit model you don't write it).

Trust isn't the issue here.

> Regarding your complaints about contiguous memory from the initial task - I don't see how this is an issue. You should have full access to all the untyped and can recycle them at will; those frames from the initial task _should_ be derived from an untyped.
> 
> I say _should_ here because I need to check the code; for some reason I don't recall anything where we create untyped with an existing watermark for those initial caps. If that doesn't exist I would very much agree with you; those definitely should exist so that the raw memory backing the initial task can be recycled by revoking (deleting?) those caps.
> 
> It's possible this is what you meant by:
> 
> > As you have frame caps, they still can't be used to create kernel objects, so it's effectively still device memory.
> 
> ... in which case, yes, I believe seL4 should be modified to fix that. Having those frame caps should still mean the untypeds still exist.

Someone had a patch to reclaim initial task's memory, but I don't think it's merged and not sure where to find it.

**Comment by Ivan-Velickovic** [comment]
@Indanz this discussion happened while I was away, sorry for late response. I agree that both approaches are somewhat hacky and a more flexibility on the kernel side would be better. Do you have anything in mind? It might be better to discuss this with @lsf37 as well, maybe at one of the developer hangouts...

**Comment by Ivan-Velickovic** [comment]
Actually I just realised that removing the loader entirely like you said would get rid of this issue as well.
